File: samples/python/rotate_write_key.py
#!/usr/bin/env python3
"""KeyNub SDK - Python sample: take ownership of a new dongle.

A dongle ships holding KeyNub's write-auth key. This replaces it with yours, so
that from the next session onward only your key can write records, erase them or
increment counters. Run it once per dongle, when it arrives.

Two keys are involved and they are not interchangeable:

  * the **current** key -- KeyNub's, supplied with the delivery. It proves you
    may replace what is there. After this runs it no longer works on this unit.
  * the **replacement** key -- yours, generated below or brought from your own
    key management. Its private half is what will authorise every future write,
    so it is worth exactly as much as your licence-signing key.

Both are P-256 private keys in PKCS#8 DER.

Targets real hardware; prints guidance and exits 0 when no dongle is attached.

    python rotate_write_key.py --current ../../keys/keynub-shipping-writeauth.key.der --new my-key.der
    python rotate_write_key.py --current ../../keys/keynub-shipping-writeauth.key.der --generate my-key.der
"""

import argparse
import sys
from pathlib import Path

try:
    import keynub_licdongle as kn
except ModuleNotFoundError:
    # Running from a checkout, where the package is not installed.
    import os
    sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                    "..", "..", "bindings", "python"))
    import keynub_licdongle as kn


def generate_key(path: Path) -> bytes:
    """A fresh P-256 key, written as PKCS#8 DER.

    Only here so the sample runs end to end. In production this belongs wherever
    you already keep signing keys -- an HSM, a key vault, whatever holds the key
    you sign licences with -- because losing it means no further writes to any
    dongle rotated to it, and there is no recovery path that does not involve
    the units coming back.
    """
    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric import ec

    if path.exists():
        raise SystemExit(f"{path} already exists; refusing to overwrite a key")
    key = ec.generate_private_key(ec.SECP256R1())
    der = key.private_bytes(serialization.Encoding.DER,
                            serialization.PrivateFormat.PKCS8,
                            serialization.NoEncryption())
    path.write_bytes(der)
    print(f"wrote a new P-256 key to {path}")
    print("Back it up before going further. Nothing below can be undone.")
    return der


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--current", required=True, type=Path,
                    help="the key the dongle holds now (KeyNub's, as delivered)")
    group = ap.add_mutually_exclusive_group(required=True)
    group.add_argument("--new", type=Path, help="your replacement key, PKCS#8 DER")
    group.add_argument("--generate", type=Path, metavar="FILE",
                       help="generate a replacement key and write it here first")
    args = ap.parse_args()

    current = args.current.read_bytes()
    replacement = generate_key(args.generate) if args.generate else args.new.read_bytes()

    with kn.LicenseDongleContext() as ctx:
        if not ctx.enumerate():
            print("Connect a KeyNub dongle and re-run.")
            return 0

        with ctx.open() as dongle:
            print(f"dongle {dongle.get_serial()}")
            with dongle.session() as session:
                # The current key proves you may replace it. Without this the
                # rotation is refused: holding the write role is the only
                # evidence the outgoing key is yours to change.
                session.authorize_write(current)

                # The replacement proves itself, by signing this session. That is
                # what makes it impossible to rotate to a key nobody holds --
                # the mistake here that no later session could undo.
                session.rotate_write_key(replacement)
                print("rotated: this dongle now answers only to your key")

                # Prove it, rather than trusting the return code. A fresh session
                # is the only place the change is observable, because the session
                # above keeps the role it was already granted.
            with dongle.session() as session:
                try:
                    session.authorize_write(current)
                except kn.LicenseDongleError:
                    print("confirmed: the old key no longer elevates")
                else:
                    print("WARNING: the old key still works -- do not ship this unit")
                    return 1
                session.authorize_write(replacement)
                print("confirmed: your key elevates")

    print("\nKeep the replacement key safe. Every future write to this dongle "
          "needs it, and it cannot be recovered from the dongle.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: samples/flat/rotate_write_key.c
/* KeyNub SDK - flat-API sample: take ownership of a new dongle.
 *
 * A dongle ships holding KeyNub's write-auth key. This replaces it with yours, so
 * that from the next session onward only your key can write records, erase them or
 * increment counters. Run it once per dongle, when it arrives.
 *
 * Both keys are P-256 private keys in PKCS#8 DER. Generate yours with:
 *
 *   openssl ecparam -name prime256v1 -genkey -noout |
 *     openssl pkcs8 -topk8 -nocrypt -outform DER -out my-key.der
 *
 * Build it with the SDK's CMake project (-DLICD_BUILD_SAMPLES=ON), or compile it
 * straight against a release archive:
 *
 *   cc rotate_write_key.c -Iinclude -Lx64 -lkeynub_licdongle_flat -o rotate_write_key
 *   ./rotate_write_key ../../keys/keynub-shipping-writeauth.key.der my-key.der
 *
 * This is the shape LabVIEW, VBA, COBOL and Fortran reproduce: integer handles,
 * caller-provided buffers, no callbacks. The key is passed as a byte array and its
 * length, which every one of those environments can express.
 *
 * Targets real hardware: with no dongle attached it prints guidance and exits 0.
 *
 * The replacement key is worth what your licence-signing key is worth. It cannot be
 * recovered from the dongle, and a unit rotated to a key you have lost has to come
 * back to be re-provisioned.
 */

#include <licd_flat.h>

#include <stdio.h>

#define MAX_KEY 4096

static void show_error(int32_t handle, const char *operation, int32_t status) {
    char text[LICDF_ERROR_SIZE] = {0};
    char detail[LICDF_ERROR_SIZE] = {0};
    licdf_strerror(status, text, (int32_t)sizeof text);
    licdf_last_error(handle, detail, (int32_t)sizeof detail);
    fprintf(stderr, "KeyNub error in %s: %s\n", operation, text);
    if (detail[0] != '\0') {
        fprintf(stderr, "  detail: %s\n", detail);
    }
}

static int32_t read_key(const char *path, uint8_t *buffer, int32_t capacity) {
    FILE *f = fopen(path, "rb");
    if (f == NULL) {
        fprintf(stderr, "cannot open %s\n", path);
        return 0;
    }
    size_t got = fread(buffer, 1, (size_t)capacity, f);
    fclose(f);
    if (got == 0) {
        fprintf(stderr, "%s is empty\n", path);
    }
    return (int32_t)got;
}

int main(int argc, char **argv) {
    if (argc != 3) {
        fprintf(stderr, "usage: %s <current-key.der> <new-key.der>\n", argv[0]);
        return 2;
    }

    uint8_t current[MAX_KEY], replacement[MAX_KEY];
    int32_t current_len = read_key(argv[1], current, (int32_t)sizeof current);
    int32_t replacement_len = read_key(argv[2], replacement, (int32_t)sizeof replacement);
    if (current_len == 0 || replacement_len == 0) {
        return 2;
    }

    int32_t count = 0;
    int32_t status = licdf_device_count(&count);
    if (status != LICD_OK) {
        show_error(0, "licdf_device_count", status);
        return 1;
    }
    if (count == 0) {
        printf("Connect a KeyNub dongle and re-run.\n");
        return 0;
    }

    int32_t handle = licdf_open(NULL); /* first dongle; pass a serial to choose */
    if (handle <= 0) {
        show_error(0, "licdf_open", handle);
        return 1;
    }

    char serial[LICDF_SERIAL_SIZE] = {0};
    if (licdf_get_serial(handle, serial, (int32_t)sizeof serial) == LICD_OK) {
        printf("dongle %s\n", serial);
    }

    int result = 1;
    if ((status = licdf_session_open(handle)) != LICD_OK) {
        show_error(handle, "licdf_session_open", status);
        goto done;
    }
    if ((status = licdf_write_auth(handle, current, current_len)) != LICD_OK) {
        show_error(handle, "licdf_write_auth", status);
        goto done;
    }
    if ((status = licdf_write_auth_rotate(handle, replacement, replacement_len)) != LICD_OK) {
        show_error(handle, "licdf_write_auth_rotate", status);
        goto done;
    }
    printf("rotated: this dongle now answers only to your key\n");
    licdf_session_close(handle);

    /* A fresh session is the only place the change is observable: the session
     * above keeps the role it was already granted. */
    if ((status = licdf_session_open(handle)) != LICD_OK) {
        show_error(handle, "licdf_session_open", status);
        goto done;
    }
    if (licdf_write_auth(handle, current, current_len) == LICD_OK) {
        fprintf(stderr, "WARNING: the old key still works -- do not ship this unit\n");
        goto done;
    }
    printf("confirmed: the old key no longer elevates\n");
    if ((status = licdf_write_auth(handle, replacement, replacement_len)) != LICD_OK) {
        show_error(handle, "licdf_write_auth", status);
        goto done;
    }
    printf("confirmed: your key elevates\n");
    licdf_session_close(handle);

    printf("\nKeep the replacement key safe. Every future write to this dongle needs it.\n");
    result = 0;

done:
    licdf_close(handle);
    return result;
}
